### PR TITLE
feat: Improve shuffle metrics (second attempt)

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

N/A

This PR replaces https://github.com/apache/datafusion-comet/pull/1173

## Rationale for this change

Changes:

- Fix: Report write time accurately (this is now just the time for writing to disk but it previously included some of the IPC encoding time)
- New: Number of input batches
- New: Encoding and compression time
- New: Total time spent in native shuffle code

### CODE REVIEW

**Comment by andygrove** [comment]
@mbutrovich @parthchandra fyi

**Comment by parthchandra** [comment]
Just for clarification - what's the relation between shuffle write time, encoding and compression total time, and native shuffle total time?
I would think `shuffle write time + encoding and compression total time = native shuffle total time`, but that does not seem to be the case?

**Comment by andygrove** [comment]
> Just for clarification - what's the relation between shuffle write time, encoding and compression total time, and native shuffle total time? I would think `shuffle write time + encoding and compression total time = native shuffle total time`, but that does not seem to be the case?

There is also evaluating the partition expressions (typically very fast if they are just column references) and then the time to actually split the batches into partitions.

**Comment by parthchandra** [comment]
> > Just for clarification - what's the relation between shuffle write time, encoding and compression total time, and native shuffle total time? I would think `shuffle write time + encoding and compression total time = native shuffle total time`, but that does not seem to be the case?
> 
> There is also evaluating the partition expressions (typically very fast if they are just column references) and then the time to actually split the batches into partitions.

From the above screenshot, `shuffle write time + encoding and compression total time = 18.4s` and  `native shuffle total time=28.8s`, so there is a difference of `10.4s` which is substantial. Wondering if we are missing something. 

Nonetheless, the PR certainly improves on the current.

**Comment by andygrove** [comment]
There is also interaction with the memory pool, which makes JNI calls into synchronized code in the JVM.

I will see if I can make the metrics more complete in this PR.

**Comment by andygrove** [comment]
@parthchandra The numbers almost add up now.

**Comment by andygrove** [comment]
Here is Gluten's equivalent for comparison:

**Comment by parthchandra** [comment]
> @parthchandra The numbers almost add up now.
> 
Brilliant!

**Review by parthchandra — APPROVED** [review]

File: docs/source/user-guide/metrics.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+| `elapsed_compute` | Total time spent in this operator, fetching batches from a JVM iterator.                            |
+| `jvm_fetch_time`  | Time spent in the JVM fetching input batches to be read by this `ScanExec` instance.                |
+| `arrow_ffi_time`  | Time spent using Arrow FFI to create Arrow batches from the memory addresses returned from the JVM. |
+
+### ShuffleWriterExec
+
+| Metric            | Description                                                   |
+| ----------------- | ------------------------------------------------------------- |
+| `elapsed_compute` | Total time excluding any child operators.                     |
+| `ipc_time`        | Time to encode batches in IPC format and compress using ZSTD. |
```
Comment (mbutrovich) [inline docs/source/user-guide/metrics.md:61]: Should we add `mempool_time` and `report_time`?

**Review by viirya — APPROVED** [review]

**Pull request merged**